pkg/prometheus/promcfg.go

```diff
@@ -83,8 +87,25 @@ func generateConfig(p *v1alpha1.Prometheus, mons map[string]*v1alpha1.ServiceMon
 
 	var scrapeConfigs []yaml.MapSlice
 	for _, identifier := range identifiers {
-		for i, ep := range mons[identifier].Spec.Endpoints {
-			scrapeConfigs = append(scrapeConfigs, generateServiceMonitorConfig(mons[identifier], ep, i, basicAuthSecrets))
+		if len(mons[identifier].Spec.Endpoints) == 0 {
+			listOpts := metav1.ListOptions{
+				LabelSelector: metav1.FormatLabelSelector(&mons[identifier].Spec.Selector),
+			}
+			svcList, err := kclient.Core().Services(mons[identifier].Namespace).List(listOpts)
```

Comment:
So far all config generation does not do any network calls or retrieval but is a simple function of input -> output, it would be good if we can keep it that way. Basically just move the retrieval part out of here and just pass in the list of services, you need.

Additionally actually calling out to the Kubernetes API is not necessary here, you can just read all Services from the informer cache of the Operator struct.